Add support for separate timeout values per connection

attilahorvath opened the pull request:
As referenced in an old issue (https://github.com/rails-sqlserver/tiny_tds/issues/84), TinyTDS currently only supports a single, global timeout value because of a limitation in FreeTDS. However, support for per-connection timeout values has been added to FreeTDS in the meantime (https://github.com/FreeTDS/freetds/commit/33fa016ca933ea5534007bce69c10e657e92e653), so we can take advantage of it now.

For some reason the call to FreeTDS expects a string value instead of an integer, so I'm forcing the conversion. This way it remains backward compatible with existing code.

metaskills commented on lib/tiny_tds/client.rb:
```diff
@@ -48,8 +48,10 @@ def initialize(opts = {})
+      forced_integer_keys = [:login_timeout, :port]
       forced_integer_keys.each { |k| opts[k] = opts[k].to_i if opts[k] }
+      forced_string_keys = [:timeout]
+      forced_string_keys.each { |k| opts[k] = opts[k].to_s if opts[k] }
```
Given that we likely continue to want this to be integers and passed down to `dbsettime(NUM2INT(timeout))` I would suggest removing this and doing the string coercion in client.c within the `dbsetopt` call.

attilahorvath commented on ext/tiny_tds/client.c:
```diff
@@ -326,8 +326,6 @@ static VALUE rb_tinytds_connect(VALUE self, VALUE opts) {
   if (!NIL_P(ltimeout))
     dbsetlogintime(NUM2INT(ltimeout));
-  if (!NIL_P(timeout))
-    dbsettime(NUM2INT(timeout));
```
I've fixed this behaviour in FreeTDS in the meantime: https://github.com/FreeTDS/freetds/pull/107. So from now it'll be possible to just simply call `dbsettime(timeout)` each time.

We should still use this original solution though to be compatible with existing versions.

metaskills commented:
@attilahorvath Sorry for the delay, I'd like to pull this in. In order to help me and other understand the final proposed changes – could you include a change to the README section `#tinytdsclient-usage` please?

attilahorvath commented:
Added a note to the `README` to explain that separate timeouts per-client won't work with FreeTDS <1.0rc5.

With newer versions, we can simply call the per-client `dbsetopt` function, so they will be all separate. However, with older versions all previously set timeout values will be overwritten each time a new client is created, since we have to fall back to the global `dbsettime` call. 

Also, I've rebased and squashed my commits to keep the history clean.

attilahorvath commented:
I don't really know why the AppVeyor test failed, it was succeeding with the same code before I rebased my branch.

Is there any way to retry it? I couldn't find a way to do that.

metaskills commented:
> Is there any way to retry it? I couldn't find a way to do that.

Since Appveyor is a paid service, they do not expose that rebuild without being logged in. I just hit it for you and will watch this today.

attilahorvath commented:
Thanks, it seems to have passed on the second try.

metaskills commented:
Thanks so much for this work! I've merged it in and it will be out in v2.1.0.pre3 a bit later today.
